zero_bin/prover/src/lib.rs

```diff
@@ -96,20 +98,31 @@ impl BlockProverInput {
             save_inputs_on_error,
         };
 
-        // Segment the batches, prove segments and aggregate them to resulting batch
-        // proofs.
-        let batch_proof_futs: FuturesUnordered<_> = block_generation_inputs
+        let vec_segment_agg_proofs: FuturesUnordered<_> = block_generation_inputs
             .iter()
-            .enumerate()
-            .map(|(idx, txn_batch)| {
+            .map(|txn_batch| {
                 let segment_data_iterator = SegmentDataIterator::<proof_gen::types::Field>::new(
                     txn_batch,
                     Some(max_cpu_len_log),
                 );
 
+                // Map over the indexed stream with the given seg_prove_ops operation.
                 Directive::map(IndexedStream::from(segment_data_iterator), &seg_prove_ops)
-                    .fold(&seg_agg_ops)
-                    .run(&runtime)
+                    .run(&proof_runtime.segment_proof_runtime)
+            })
+            .collect();
+
+        // Await all the futures concurrently
+        let vec_segment_agg_proofs: Vec<_> = vec_segment_agg_proofs.collect().await;
```

Comment:
So here, you first await for all the segment proofs for all generation inputs from all the blocks (this function is called  in parallel for every block from the input interval). This is certainly not the right way. You will need zillion bytes of memory on `zero_bin` computer to keep all this non aggregated proofs.

I think we should keep the old implementation. There, you chain the operations `seg_prove_ops` and `seg_agg_ops` (future inside of the future), and you get one future for every batch aggregated proof. So locally, every `BatchAggregatableProof` became result of one async task executed in parallel. But in reality (for non in-memory execution), every `seg_prove_ops` and `seg_agg_ops` is sent to the remote worker machine. So it seems to me that  `paladin` should be dealing with correct scheduling of operations to appropriate worker server depending how hungry the operation is, no need to change anything in the `zero_bin` implementation.

Reply:
Maybe I am missing something, but it feels to me that this task goes down to:
If paladin gets the `ops::SegmentProof`, it should send it to strong remote worker machine.
If paladin gets the `ops::SegmentAggProof` or `ops::BatchAggProof`, it should send it to the cheaper remote worker machine.

Reply:
https://github.com/0xPolygonZero/zk_evm/issues/627

Created this task for handling this in future.